### PR TITLE
CRI validation test

### LINKED ISSUE #34040: CRI validation test
This was brought in the @kubernetes/sig-node meeting previously. I think having a CRI validation test is necessary eventually to qualify a new CRI shim and help CRI shim developers. It would be good to have such a test sooner than later. 

Basically, we want a test framework and a set of tests to validate a CRI (grpc) server alone. For starters, we can just run tests locally on a machine where the grpc server runs. The test only needs to know the socket to talk to, and can perform the basic operations such as creating/deleting sandboxes/containers. Ideally, no runtime-specific bits should be included in the tests. 

The CRI validation test should replace/subsume the existing runtime validation test (formerly known as automatic docker validation test: https://github.com/kubernetes/features/issues/57).  

@kubernetes/sig-node, since we have multiple ongoing projects implementing CRI shims, this should be a sig-node community effort. If someone already has testing bits in their repository, you're welcome to donate such tests and/or harness. Volunteers to work on the CRI validation test is also welcome.

/cc @dchen1107 @Random-Liu @euank @yifan-gu @mrunalp

### PR DESCRIPTION
**Pull request by xlgao-zju**
This pr use e2e framework to validate a CRI (grpc) server alone. I test this pr against dockershim and frakti. It works fine.

**How to run cri e2e test**

``` sh
cd $GOPATH/src/k8s.io/kubernetes
make test-e2e-cri
```

The default server is dockershim. If we want to test other CRI server such as frakti, we can `export CONTAINER_RUNTIME_ENDPOINT=/var/run/frakti.sock`. And we can run this test against frakti. 

And we also can `go get`  this package, then run the following commands to run this test.

``` sh
go run test/e2e_cri/runner/run.go
```

If we don't need to build dependencies, we can add this flag `--build-dependencies=false`.

A beginning of #34040

/cc @Random-Liu @yujuhong @resouer @feiskyer 

Signed-off-by: Xianglin Gao xlgao@zju.edu.cn

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35266)

### CODE REVIEW

File: test/e2e/framework/framework.go
Comment (feiskyer) [inline test/e2e/framework/framework.go]: nit: replace `Cri` with `CRI` (also other places)

Comment (xlgao-zju) [inline test/e2e/framework/framework.go]: @feiskyer updated

**Comment by feiskyer** [comment]
@xlgao-zju Could this test package be installed by `go get` (e.g. `go get -u k8s.io/kuernetes/test/e2e_cri`)? This could make the validation procedure much easier for runtime maintainers.

**Comment by xlgao-zju** [comment]
It can't be installed by `go get` for now. I will work on it.

**Comment by xlgao-zju** [comment]
@feiskyer Updated. We can run this command to run CRI e2e test for now.
```sh
go run test/e2e_cri/runner/run.go --server-address=/var/run/frakti.sock
```
If we don't need to build dependencies, we can add this flag `--build-dependencies=false`.

File: test/e2e/framework/framework.go
Comment (resouer) [inline test/e2e/framework/framework.go]: nit: extra spaces

**Comment by pskrzyns** [comment]
@xlgao-zju I dont think it should close #34040 cause this missing i.e. pod tests (blocked by other PR).
I think we should use #34040 as umbrella issue for e2e CRI testing (maybe tracking implemented/unimplemented tests).

What about stopping/removing/listing sandbox/image which doesn't exist ?

File: test/e2e/framework/framework.go
Comment (xlgao-zju) [inline test/e2e/framework/framework.go]: Cause this line is in the `if !f.cri{}`, so I add a tab.

**Comment by xlgao-zju** [comment]
@pskrzyns Sounds reasonable. This pr only implements the framework, and it is the beginning of #34040. I will use #34040 to track other related prs in the future. And the test cases in this pr actually prove that the framework works fine. So, I will add more test cases in following prs.

**Comment by feiskyer** [comment]
@xlgao-zju nit: #33988 has been in.

**Comment by xlgao-zju** [comment]
@feiskyer will run this pr via dockershim and add container related test cases.

**Comment by yujuhong** [comment]
@Random-Liu's #35561 will turn on grpc for dockershim by default. @xlgao-zju you should be able to test dockershim after that.

**Comment by xlgao-zju** [comment]
@feiskyer @yujuhong updated.
I test this pr against dockershim, and it works fine.

**Comment by yujuhong** [comment]
@xlgao-zju sorry for the delay. We've been busy trying to get the api/docker integration completed before the code freeze. This PR is a nice-to-have, but is not mandatory for 1.5. We'll try to get to this PR this week, but don't be surprised if it didn't happen.

**Comment by xlgao-zju** [comment]
@yujuhong get it.

File: test/e2e_cri/container.go
Comment (feiskyer) [inline test/e2e_cri/container.go]: nit:  replace internalApi -> internalapi, runtimeApi -> runtimeapi

File: test/e2e_cri/container.go
Comment (feiskyer) [inline test/e2e_cri/container.go]: CRI assumes all fields in PodSandboxMetadata are set by kubelet, maybe extract a helper for default metadata?

File: test/e2e_cri/container.go
Comment (feiskyer) [inline test/e2e_cri/container.go]: CRI assumes all fields in ContainerMetadata are set by kubelet, maybe extract a helper for default metadata?

File: test/e2e_cri/e2e_cri_test.go
Comment (feiskyer) [inline test/e2e_cri/e2e_cri_test.go]: nit: rename `more_reporters` to `morereporters`, maybe just not alias instead?

File: test/e2e_cri/image.go
Comment (feiskyer) [inline test/e2e_cri/image.go]: nit: rename to RemoveAllImages and only remove images related to the tests.

File: test/e2e_cri/image.go
Comment (feiskyer) [inline test/e2e_cri/image.go]: nit: Should add a filter here

File: test/e2e_cri/image.go
Comment (feiskyer) [inline test/e2e_cri/image.go]: same here, add a filter. maybe check RepoTags instead of counts?

File: test/e2e_cri/pod.go
Comment (feiskyer) [inline test/e2e_cri/pod.go]: nit: set all fields of PodSandboxMetadata

File: test/e2e_cri/runner/run.go
Comment (feiskyer) [inline test/e2e_cri/runner/run.go]: should also add a image-service-address, and set it same with runtime-address if not specified.

**Review by feiskyer — CHANGES_REQUESTED** [review]

**Comment by xlgao-zju** [comment]
@feiskyer updated.

File: test/e2e_cri/container.go
Comment (pskrzyns) [inline test/e2e_cri/container.go]: I think that removing sandbox should be done in AfterEach, and creation of sandbox should be in BeforeEach

Comment (xlgao-zju) [inline test/e2e_cri/container.go]: @pskrzyns Yes, it is clean. But it is not flexible in some test cases(maybe added in the future), such as a test case requires two pods.
I think these basic test cases about container will only require one pod. So, I will warp these cases with a 'Context()'. I will move removing sandbox into 'AfterEach', and creation of sandbox into 'BeforeEach' .

**Comment by pskrzyns** [comment]
@yujuhong can we merge it to branch other than master ?
It would be nice to have it in main repo but i don't think it should be master yet.

@jayunit100 PTAL

File: test/e2e_cri/container.go
Comment (xlgao-zju) [inline test/e2e_cri/container.go]: @pskrzyns Any thing else would you like to refact, maybe I could do it in my pr directly.

**Comment by xlgao-zju** [comment]
> @yujuhong can we merge it to branch other than master ?
It would be nice to have it in main repo but i don't think it should be master yet.

@pskrzyns why this pr should not be in master yet?

**Comment by feiskyer** [comment]
> @yujuhong can we merge it to branch other than master ?
> It would be nice to have it in main repo but i don't think it should be master yet.

@pskrzyns Why this? We don't merge PRs to non-master branch except cherry-picks.

**Comment by pskrzyns** [comment]
@xlgao-zju as this pr is for implementing framework I'm missing some test primitives/helpers, i.e. 
```go
func createPodSandbox(c internalapi.RuntimeService, prefix string ) (string, error) {
	By("create a PodSandbox with name")
	name := prefix + string(uuid.NewUUID())
	config := &runtimeapi.PodSandboxConfig{
		Metadata: buildPodSandboxMetadata(&name),
	}

	return c.RunPodSandbox(config)
}

func createPodSandboxOrFail(c internalapi.RuntimeService, prefix string ) string {
	podId, err := createPodSandbox(c, prefix)
	framework.ExpectNoError(err, "Failed to create PodSandbox: %v", err)
	framework.Logf("Created PodSandbox %s\n", podId)
	return podId
}

func getPodSandboxStatus(c internalapi.RuntimeService, podId string) (*runtimeapi.PodSandboxStatus, error) {
	By("get PodSandbox status")
	return c.PodSandboxStatus(podId)

}

func getPodSandboxStatusOrFail(c internalapi.RuntimeService, podId string) *runtimeapi.PodSandboxStatus {
	status, err := getPodSandboxStatus(c, podId)
	framework.ExpectNoError(err, "Failed to get PodSandbox %s status: %v", podId, err)
	return status
}

func verifyPodSandboxStatus(c internalapi.RuntimeService, podId string, expectedStatus *runtimeapi.PodSandboxStatus, statusName string) {
	status := getPodSandboxStatusOrFail(c, podId)
	Expect(*status.State).To(Equal(expectedStatus), "PodSandbox state should be " + statusName)
}

func testCreatePodSandbox(c internalapi.RuntimeService) string {
	podId := createPodSandboxOrFail(c, "PodSandbox-for-create-test-")
	verifyPodSandboxStatus(c, podId, runtimeapi.PodSandboxState_SANDBOX_READY, "ready")
	return podId
}

func stopPodSandbox(c internalapi.RuntimeService, podID string) (error) {
	By("stop pod PodSandbox with podId")
	return c.StopPodSandbox(podID)
}

func removePodSandbox(c internalapi.RuntimeService, podID string) (error) {
	By("remove pod PodSandbox with podId")
	return c.RemovePodSandbox(podID)
}

func stopPodSandboxOrFail(c internalapi.RuntimeService, podID string) {
	err := stopPodSandbox(c, podID)
	framework.ExpectNoError(err, "Failed to stop PodSandbox: %v", err)
	framework.Logf("Stopped PodSandbox %s\n", podID)
}

func removePodSandboxOrFail(c internalapi.RuntimeService, podID string) {
	err := removePodSandbox(c, podID)
	framework.ExpectNoError(err, "Failed to remove PodSandbox: %v", err)
	framework.Logf("Removed PodSandbox %s\n", podID)
}
```

I think they should be implemented in ```util.go``` so they can be used in both ```pod.go``` and ```container.go`` files, cause at this moment there is a lot of code duplication in those tests.
Sorry for nagging so much, but from my experience refactoring framework after people start using it is real pain in the neck.

**Comment by pskrzyns** [comment]
> @pskrzyns Why this? We don't merge PRs to non-master branch except cherry-picks.

@feiskyer true, I still have some problems with differences betwen github flow and git flow ;)

**Comment by resouer** [comment]
@pskrzyns I think there's misunderstand, yuju just means node team doesn't have enough bandwidth for a quick review at now. This PR is definitely still under review (and far from merge actually), so you can just start a review to post your suggestions :)

**Comment by xlgao-zju** [comment]
>I think they should be implemented in util.go so they can be used in both pod.go and `container.go files, cause at this moment there is a lot of code duplication in those tests.
Sorry for nagging so much, but from my experience refactoring framework after people start using it is real pain in the neck.

@pskrzyns Thank you for your suggestions, I would like to reduce code duplication too. So I implement some functions in `util.go`, maybe its good for the following test cases :)

File: test/e2e_cri/pod.go
Comment (pskrzyns) [inline test/e2e_cri/pod.go]: should we use ```ListPodSandbox```  or ```PodSandboxStatus``` with speciffic error to validate if pod doesn't exist ? @yujuhong

File: test/e2e_cri/util.go
Comment (pskrzyns) [inline test/e2e_cri/util.go]: it's not

File: test/e2e_cri/util.go
Comment (pskrzyns) [inline test/e2e_cri/util.go]: empty line

File: test/e2e_cri/util.go
Comment (pskrzyns) [inline test/e2e_cri/util.go]: move timeout to constant

File: test/e2e_cri/container.go
Comment (pskrzyns) [inline test/e2e_cri/container.go]: This is one big test case, if ```By("test start container")``` fails it won't check if removing container works correctly.
It should be split to five independent test cases

File: test/e2e/framework/framework.go
Comment (Random-Liu) [inline test/e2e/framework/framework.go]: Why you need e2e framework for cri validation test? I think we should add another util or framework.

File: test/e2e/framework/test_context.go
Comment (Random-Liu) [inline test/e2e/framework/test_context.go]: ditto, no need to change e2e framework.

File: test/e2e/framework/util.go
Comment (Random-Liu) [inline test/e2e/framework/util.go]: ditto.

File: test/e2e/framework/framework.go
Comment (xlgao-zju) [inline test/e2e/framework/framework.go]: @Random-Liu True, CRI validation test is not associated with existing e2e test framework. So should we add another independent framework(which is specially for CRI validation test) like e2e framework? And should we put these files in `/test` or in `/kubelet`?

File: test/e2e_cri/container.go
Comment (xlgao-zju) [inline test/e2e_cri/container.go]: @pskrzyns Thank you for your suggestions.  I will fix these in next update. :)

File: test/e2e/framework/framework.go
Comment (Random-Liu) [inline test/e2e/framework/framework.go]: I think put it in `test/e2e_cri` is fine. :)

**Comment by xlgao-zju** [comment]
@Random-Liu updated.

File: test/e2e_cri/container.go
Comment (xlgao-zju) [inline test/e2e_cri/container.go]: @Crazykev Since we don't clean up images after testing, so we don't need to pull the image before each test case.

Comment (Crazykev) [inline test/e2e_cri/container.go]: @xlgao-zju Frakti(or HyperContainer) won't pull image at container creation time, if that image doesn't exist, there will be an error, so for safety maybe we still need a pull here.

File: test/e2e/framework/pods.go
Comment (jawnsy) [inline test/e2e/framework/pods.go]: I think the previous form "[Convenience method](http://wiki.c2.com/?ConvenienceMethods)" was better, but that's just one opinion :)

File: test/e2e/kubectl.go
Comment (jawnsy) [inline test/e2e/kubectl.go]: Reustable -> Reusable as well

File: test/e2e/framework/pods.go
Comment (xlgao-zju) [inline test/e2e/framework/pods.go]: Will revert this.

File: test/e2e/kubectl.go
Comment (xlgao-zju) [inline test/e2e/kubectl.go]: ACK

**Comment by feiskyer** [comment]
@xlgao-zju We should close this PR and move on to https://github.com/kubernetes-incubator/cri-tools.

**Comment by xlgao-zju** [comment]
@feiskyer Yes, we can close this pr now.

**Pull request closed without merging**
